<?xml version="1.0" ?>
<launch>
    <!--ADD ONS PARAMETERS -->
    <arg name="use_rviz_gui" default="false"/>
    <arg name="use_gazebo_gui" default="false"/>

    <!-- ROBOT PARAMETERS -->
    <param name="robotName" value="kuka"/>
    <param name="groupName" value="kuka_arm"/>
    <param name="collisionBaseLinksNames" type="yaml" value="[base_link]"/>

    <arg name="initial_configuration_name" value="START0"/>
    <param name="initialConfigurationName" value="$(arg initial_configuration_name)"/>  
    
    <arg name="kinematics_solver" default="bio_ik/BioIKKinematicsPlugin"/>

    <arg name="planner_name" default="BiTRRT"/>
    <param name="plannerName" value="$(arg planner_name)"/>
    <arg name="planner_objective" default="PathLengthOptimizationObjective"/>
    <param name="plannerObjective" value="$(arg planner_objective)"/>

    <!-- ROBOT SAFETY PARAMETERS -->
    <arg name="safety_distance" default="0.03"/>

    <!-- ROBOT TOOL PARAMETERS -->
    <arg name="tool_name" default="none"/>
    <param name="toolName" value="$(arg tool_name)"/>

    <arg name="end_effector_name" value="kuka_$(arg tool_name)" unless="$(eval arg('tool_name') == 'none' or arg('tool_name') == 'gripper')"/> 
    <arg name="end_effector_name" value="tool0" if="$(eval arg('tool_name') == 'none')"/> 
    <param name="endEffectorName" value="$(arg end_effector_name)"/>

    <!-- ROBOT CALIBRATION PARAMETERS -->
    <arg name="calibration" default="false"/>
    <param name="calibration" value="$(arg calibration)"/>

    <arg name="calibrated_model_file" default="" unless="$(arg calibration)"/> 
    <arg name="calibrated_model_file" default="TODO" if="$(arg calibration)"/>

    <rosparam file="$(arg calibrated_model_file)" command="load" if="$(arg calibration)"/>

    <!-- EXECUTION PARAMETERS -->
    <node pkg="robot_arm_tools" name="pause_button" type="pause_button"/>
    <arg name="simulation" default="true"/>
    <param name="simulation" value="$(arg simulation)"/>
    <arg name="robot_ip" default="172.31.1.147" />  
    <param name="robotIP" value="$(arg robot_ip)"/>
    <arg name="gazebo" default="false"/>

    <!-- LOCALISATION PARAMETERS -->
    <!--TODO Take into account an eventual robot base offset ? -->
    <!--<node pkg="tf2_ros" type="static_transform_publisher" name="virtual_joint_broadcaster_0" args="0 0 0 0 0 0 world base_link" />-->
    <node pkg="tf2_ros" type="static_transform_publisher" name="virtual_joint_broadcaster_1" args="0 0 0 0 0 0 world map" /> 
    
    <!-- LAUNCH INCLUDES -->

    <!-- MoveIt based simulation -->
    <group if="$(eval arg('simulation') and not arg('gazebo'))">
        <!--To adapt !-->
        <include file="$(find kuka_kr10r1100sixx_moveit_config)/launch/demo.launch" pass_all_args="true">
            <arg name="use_rviz" value="$(arg use_rviz_gui)"/>
        </include>

        <param name="/move_group/initial" type="yaml" value="[{group: kuka_arm, pose: $(arg initial_configuration_name)}]"/>
    </group>

    <!-- Real hardware execution -->
    <group unless="$(arg simulation)">
        <!--To adapt !-->
        <include file="$(find kuka_kr10r1100sixx_moveit_config)/launch/moveit_planning_execution_eki.launch">
            <arg name="sim" value="false"/>
            <arg name="robot_ip" default="$(arg robot_ip)" />
        </include>

        <node name="RViz" pkg="rviz" type="rviz" if="$(arg use_rviz_gui)"/>
    </group>

    <!-- Gazebo based simulation -->
    <group if="$(eval arg('simulation') and arg('gazebo'))">
        <!--To adapt or comment if there is no Gazebo support !-->
        <include file="$(find kuka_kr10r1100sixx_moveit_config)/launch/gazebo.launch" pass_all_args="true">
            <arg name="gazebo_gui" value="$(arg use_gazebo_gui)"/>
        </include>

        <remap from="/scaled_pos_joint_traj_controller/follow_joint_trajectory" to="/pos_joint_traj_controller/follow_joint_trajectory"/>

        <node name="RViz" pkg="rviz" type="rviz" if="$(arg use_rviz_gui)"/>
    </group>

    <!-- Weird kuka stuff -->
    <param name="move_group/trajectory_execution/execution_duration_monitoring" value="false"/>
    <rosparam param="eki/socket_timeout">20</rosparam>

    <!-- Overwritting kinematics plugin for calibration -->
    <!--<param name="robot_description_kinematics/kuka_arm/kinematics_solver" value="robot_kinematics_plugin" if="$(arg calibration)"/>-->
    <param name="robot_description_kinematics/kuka_arm/kinematics_solver" value="$(arg kinematics_solver)"/>

    <param name="robot_description_kinematics/kuka_arm/mode" value="gd_c_8"/>

    <param name="robot_description_kinematics/kuka_arm/kinematics_solver_search_resolution" value="0.00001"/>  <!--Not used for bio_ik-->
    <param name="robot_description_kinematics/kuka_arm/dpos" value="0.00001"/>  <!--Used for bio_ik-->
    <param name="robot_description_kinematics/kuka_arm/drot" value="0.00001"/>  <!--Used for bio_ik-->
    <param name="robot_description_kinematics/kuka_arm/dtwist" value="0.00001"/>  <!--Used for bio_ik-->

    <param name="robot_description_kinematics/kuka_arm/kinematics_solver_timeout" value="1.0"/>

    <!-- Overwritting robot_description and robot_description_semantic -->
    <param name="robot_description" command="/bin/bash $(find robot_arm_tools)/scripts/GetRobotDescription.sh 'robot_arm_tools Setup_kuka.launch simulation:=$(arg simulation) gazebo:=$(arg gazebo)' '$(find robot_arm_tools)/config/tools/$(arg tool_name)/$(arg tool_name).urdf.xacro robot_name:=kuka safety_distance:=$(arg safety_distance)' '$(arg calibrated_model_file)'" unless="$(eval (arg('tool_name') == 'none' and not arg('calibration')) or arg('tool_name') == 'gripper')"/>
    <param name="robot_description_semantic" command="/bin/bash $(find robot_arm_tools)/scripts/GetRobotDescriptionSemantic.sh 'robot_arm_tools Setup_kuka.launch simulation:=$(arg simulation) gazebo:=$(arg gazebo)' '$(find robot_arm_tools)/config/tools/$(arg tool_name)/$(arg tool_name).urdf.xacro robot_name:=kuka safety_distance:=$(arg safety_distance)' '$(arg end_effector_name)' '$(arg calibrated_model_file)'" unless="$(eval (arg('tool_name') == 'none' and not arg('calibration')) or arg('tool_name') == 'gripper')"/>

</launch>
